tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc

```diff
@@ -947,9 +947,60 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
   return success();
 }
 
-// TODO(hinsu): Add canonicalization pattern for ShapeN ops that don't have all
+namespace {
+// Canonicalization pattern for ShapeNOp that don't have all
 // static input shapes. Replacing output values corresponding to static input
 // types may enable optimizations in users of the values.
+class ShapeNPartialStaticInputShape : public OpRewritePattern<ShapeNOp> {
+  using OpRewritePattern<ShapeNOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(ShapeNOp op,
+                                PatternRewriter &rewriter) const override {
+    // ShapeNOp::fold handles this case.
+    if (op.getNumOperands() == 0) return success();
+    int width = op.getType(0)
+                    .cast<ShapedType>()
+                    .getElementType()
+                    .getIntOrFloatBitWidth();
+
+    SmallVector<Value, 4> results(op.getNumOperands());
+    SmallVector<int64_t, 4> dynamic_index;
+    SmallVector<Value, 4> dynamic_input;
+    SmallVector<Type, 4> result_types;
+    for (auto e : llvm::enumerate(op.getOperands())) {
+      if (Attribute result = ConvertShapeToAttr(e.value().getType(), width)) {
+        results[e.index()] =
+            rewriter.create<TF::ConstOp>(op.getLoc(), result);
+      } else {
+        dynamic_index.push_back(e.index());
+        dynamic_input.push_back(e.value());
+        result_types.push_back(op.getType(e.index()));
+      }
+    }
+
+    if (dynamic_input.size() == op.getNumOperands()) {
+      // Cannot canonicalize ShapeN if all inputs are dynamic.
+      return failure();
+    }
+
+    if (!dynamic_input.empty()) {
```

Comment:
Now that we handle all static shaped operands case here, we can drop the folder.